### PR TITLE
Add fallback to Drastic BIOS and fake firmware when not provided

### PR DESCRIPTION
**Pull request by Swordfish90**
Hi guys, this pull request makes BIOSes and firmware files optional in MelonDS.

I've tried to get these approved upstream but the two pull requests have been waiting for a few months now:
* https://github.com/Arisotura/melonDS/pull/787
* https://github.com/Arisotura/melonDS/pull/778

It contains:
* The two free BIOS replacements which are bundled into Drastic (released with a permissive license you find in the commit)
* A fake firmware that allows setting some custom settings
* The libretro plumbing code

This has been battle-tested in Lemuroid (where you can't even set a custom BIOS file) in production since October and compatibility has been great. It might make sense to squash the commit to avoid conflicts if this gets merged upstream at a later time.

Feel free to let me know if there is anything you'd like me to change or improve.

### CODE REVIEW

**Comment by twinaphex** [comment]
Sounds great, I definitely think we should strongly consider this. What do you think @m4tsa ?

**Comment by m4tsa** [comment]
> 
> 
> Sounds great, I definitely think we should strongly consider this. What do you think @m4tsa ?

It would most likely be merged upstream if you just remind them of the PR, but I guess it could be merged and then reverted if it was ever merged upstream.

Also how much testing have you done to ensure there are no major bugs? If there are any it should probably do a warning notification saying it did fallback to freebios.

**Comment by Swordfish90** [comment]
> It would most likely be merged upstream if you just remind them of the PR, but I guess it could be merged and then reverted if it was ever merged upstream.

Hi @m4tsa , I've tried a bunch of times, while my other pull requests were approved very soon, these ones are still waiting.

> Also how much testing have you done to ensure there are no major bugs? If the compatibility is not 100% it should probably do a warning notification saying it did fallback to freebios.

Last year I did this work and personally tested it on ~50 games. I later pushed it into production with Lemuroid (~30k active users, with DS being one of the most used platforms). The BIOS files themselves are used by default on Drastic, which has more than 1M installs on Google Play. There is currently a line in the log, but if you prefer I can also forward the message to the frontend.

**Comment by m4tsa** [comment]
> Last year I did this work and personally tested it on ~50 games. I later pushed it into production with Lemuroid (~30k active users, with DS being one of the most used platforms). The BIOS files themselves are used by default on Drastic, which has more than 1M installs on Google Play. There is currently a line in the log, but if you prefer I can also forward the message to the frontend.

It should only be needed with a visual warning if there are any known incompatibilities

File: freebios/Makefile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+TC_PREFIX  = /home/exophase/pandora-dev
+PREFIX     = $(TC_PREFIX)/arm-2011.03
+AS         = $(PREFIX)/bin/arm-none-linux-gnueabi-gcc
+OBJCOPY    = $(PREFIX)/bin/arm-none-linux-gnueabi-objcopy
```
Comment (m4tsa) [inline freebios/Makefile:4]: This should probably be changed so you can actually build it with any newer arm crossbuilding toolchain

Comment (Swordfish90) [reply freebios/Makefile:4]: The source code and Makefile come directly from the original source. I just converted the blobs into the byte arrays there. I guess you're right we could try rebuilding the BIOS files and updating the Makefile. I'll work on it in the next days.

**Comment by Swordfish90** [comment]
>It should only be needed with a visual warning if there are any known incompatibilities

Ok, I'd guess we don't need it then. I'm not aware of any incompatibilities nor I've received any complaints at the moment. IMHO, adding it later if some issues arise is the best bet.

**Comment by twinaphex** [comment]
Hi there @Swordfish90 , I am told we still have an outstanding issue with the Makefile that needs resolving. Could you do that please?

**Comment by Swordfish90** [comment]
Hi @twinaphex, sorry for the late update but it has been a couple of crazy weeks and I wanted to better take a look into this.

I've been able to build the bios files from source using a more recent arm toolkit (gcc-arm-10.2-2020.11) and they seem to work correctly in a few games I tested.

The main issue here is that the resulting file has 21 different bytes compared to the original blob, so I'm not sure on how we should proceed. @m4tsa what do you think would be the best course of action here?

**Comment by m4tsa** [comment]
> Hi @twinaphex, sorry for the late update but it has been a couple of crazy weeks and I wanted to better take a look into this.
> 
> I've been able to build the bios files from source using a more recent arm toolkit (gcc-arm-10.2-2020.11) and they seem to work correctly in a few games I tested.
> 
> The main issue here is that the resulting file has 21 different bytes compared to the original blob, so I'm not sure on how we should proceed. @m4tsa what do you think would be the best course of action here?

Depends on what the bytes are used for, but still it is probably good enough for a merge.

**Comment by Swordfish90** [comment]
> Depends on what the bytes are used for, but still it is probably good enough for a merge.

Indeed :sweat_smile:. The problem is I'm not sure if these changes are harmless or not, all the tests were done with the included binaries. We can either keep them and merge it as it is, or replace them with the new ones and update the pull request.

 Just let me know which works best for you and I'll follow.

**Comment by m4tsa** [comment]
> > Depends on what the bytes are used for, but still it is probably good enough for a merge.
> 
> Indeed sweat_smile. The problem is I'm not sure if these changes are harmless or not, all the tests were done with the included binaries. We can either keep them and merge it as it is, or replace them with the new ones and update the pull request.
> 
> Just let me know which works best for you and I'll follow.

Just update the Makefile and keep the pre-compiled BIOS for now until it can be investigated or regression tested.

**Comment by Swordfish90** [comment]
Hi @m4tsa I pushed the updated makefile, just feel free to let me know if you need anything else on this pull request.

**Comment by m4tsa** [comment]
Looks fine and it compiled fine locally so I am just going to merge it.

**Pull request merged**